src/tool/confiscan.sh

```diff
@@ -115,6 +115,68 @@ else
     mkdir -p "${output_dir}"
 fi
 
+####################
+# Users and groups #
+####################
+
+passwd_file="/etc/passwd"
+[[ -f "${passwd_file}" ]] || error "passwd file not found." 1
+[[ -r "${passwd_file}" ]] || error "passwd file not readable." 1
+
+groups_file="/etc/group"
+[[ -f "${groups_file}" ]] || error "group file not found." 1
+[[ -r "${groups_file}" ]] || error "group file not readable." 1
+
+[[ -d "${output_dir}/users" ]] || mkdir -p "${output_dir}/users"
+[[ -d "${output_dir}/groups" ]] || mkdir -p "${output_dir}/groups"
+
+printf 'Username,UID,GID,Shell,Home\n' > "${output_dir}/users/users.csv"
+cut -d: -f1 ${passwd_file} | paste -d, - \
+    <(cut -d: -f3 ${passwd_file}) \
+    <(cut -d: -f4 ${passwd_file}) \
+    <(cut -d: -f7 ${passwd_file}) \
+    <(cut -d: -f6 ${passwd_file}) >> \
+    "${output_dir}/users/users.csv"
+
+info "Users found on the system:"
+column -t -s, "${output_dir}/users/users.csv"
+
+info "Users with login shells:"
+while read -r shell; do
+    [[ "${shell}" =~ ^#|^$ ]] && continue
+
+    info "Users with '${shell}':"
+    # ||: is required since the script will exit with 1
+    # if no users are found with the shell.
+    grep -E ":${shell}$" ${passwd_file} | cut -d: -f1 || :
+done < "/etc/shells"
+
+info "Privileged users:"
+while IFS=':' read -r group _ _ user; do
+    case "${group}" in
+        wheel|sudo)
+            printf "%s\n" "${user}"
+    esac
+done < "${groups_file}"
+
+info "Custom sudo rules:"
+printf 'Username,Privilege,Source\n' > "${output_dir}/users/sudoers.csv"
+while read -r user priv; do
+    printf "%s,%s,%s\n" "${user##*:}" "${priv}" "${user%%:*}" >> \
+        "${output_dir}/users/sudoers.csv"
+done < <(grep -E '^[^#]*ALL=' /etc/sudoers /etc/sudoers.d/*)
```

Comment:
I know you would like to restrict yourself to checking files only, but you are really missing potential rules here.

sudo -U bdeferme -l:

```
User bdeferme may run the following commands on lx-20202034:
    (ALL) NOPASSWD: ALL
    (pkg-build) NOPASSWD: /usr/share/lpf/scripts/lpf-pkgbuild
    (pkg-build) NOPASSWD: /usr/bin/kill
    (root) NOPASSWD: /usr/share/lpf/scripts/lpf-kill-pgroup
    (root) NOPASSWD: /usr/share/lpf/scripts/lpf-sudo-builddep
    (root) /usr/share/lpf/scripts/lpf-sudo-install
```

Your script:

```
1   │ Username,Privilege,Source
2   │ root,ALL=(ALL)  ALL,/etc/sudoers
3   │ %wheel,ALL=(ALL)    NOPASSWD: ALL,/etc/sudoers
```

Reply:
I'm able to get the rules for each user and convert them to CSV:

```bash
info "Custom sudo rules for each user:"
printf 'Username,Privilege\n' > "${output_dir}/users/sudoers.csv"
while read -r user; do
    while read -r priv; do
        printf "%s,%s\n" "${user}" "${priv}" >> \
            "${output_dir}/users/sudoers.csv"
    done < <(sudo -U "${user}" -l | sed -n '/User/,$p' | tail -n +2 | sed -e 's/    //g')
done < <(cut -d: -f1 ${passwd_file})
```

But I'm unable to get the rules for each group. Either I'm blind or it's not stated in the man-page of `sudo`. Or is this useless to also list them for groups?

Reply:
Maybe something like this:

```
grep -E '^%<group_name>' /etc/sudoers /etc/sudoers.d/*
```

for the groups would work?

Reply:
To answer the question about groups, there is no option in sudo to list them for groups, this is because groups don't really have sudo rules applied to them. A user has sudo rules applied based on the rules for the user plus the rules for the groups the user is part of.

So if you have all rules for all users listed already, you could say it's redundant to also list the rules that are found for groups.

Reply:
I've removed the code that gathers the rules for each group. Thanks!